### DISCUSSION TITLE
Tab activity indicator

### DISCUSSION THREAD

**Discussion by anongit** [wezterm/wezterm#796] [Ideas]
In iTerm2 there is a built-in option to show that there was new output in an inactive tab:

![image](https://user-images.githubusercontent.com/3994490/118294120-d778c180-b4f3-11eb-83ca-0cffb0ddd717.png)

Is there an easy way to configure this?
I guess you could make it with user vars and monitoring tabs / panes in `format-tab-title`, but it seems like a hack.

**Comment by wez** [comment]
We don't have this today, but adding support for it is doable. In my mind this overlaps a bit with https://github.com/wez/wezterm/issues/3

**Reply by anongit** [reply]
Thanks, subscribed to that issue.

**Comment by hoverduck** [comment]
I hacked this together with uservars and a bash script that you pipe your log output into.  So it won't work as something you could turn on ad-hoc for an existing long-running process, but for my use case of keeping a tail on a log file it works great.

It works with `tail -f somefile.log | wezterm-tail-alert.sh`, which just reads and echos each line, and both prints out a UserVar named `lastOutput` with the current timestamp and rings the system bell.  For any tab with `lastOutput` set, the `format-tab-title` event keeps track of the last time I had that as my active tab, and if the timestamp of `lastOutput` is greater than `lastVisit` it adds `(!)` to the pane title and turns it red.

Separately in the `bell` event I'm tracking the last line of output on those tabs, so I can ignore any log output that matches a given string (basically in my use case I want to ignore Info output and only be alerted on Warning/Fatal).  I could also imagine building off of that to change the tab color for warnings, or send a toast notification for fatals.  But that gets really application-specific.

This is the most I've ever written in Lua, so happy to know if there's an easier way to do any of this.

wezterm-tail-alert.sh:
```bash
#!/bin/bash

while read line; do
        echo $line
        printf "\e]1337;SetUserVar=%s=%s\a" lastOutput $(date +'%s'| base64)
        tput bel
done
```

```lua
local wezterm = require 'wezterm';
local lastVisits = {};
local lastLines = {};

wezterm.on("format-tab-title", function( tab, tabs, panes, config, hover, max_width )
  local pane_title = tab.active_pane.title
  local lastOutput = tonumber( tab.active_pane.user_vars.lastOutput )
  local currTime = tonumber( wezterm.strftime("%s") )
  local index = tonumber( tab.tab_index )+1
  local format = {}

  if lastOutput ~= nil then
    if lastVisits[ tab.tab_id ] == nil then
      lastVisits[ tab.tab_id ] = currTime
    end

    if lastOutput > lastVisits[ tab.tab_id ] then
      i, j = string.find( lastLines[ tab.active_pane.pane_id ], "INFO:" )
      if nil == i then
        pane_title = "(!) " .. pane_title
	table.insert( format, { Foreground={Color="Red"} } )
      end
    end

    if tab.is_active and lastVisits[ tab.tab_id ] < currTime then
      lastVisits[ tab.tab_id ] = currTime
    end
  end

  table.insert( format, { Text=index .. ": " .. pane_title } )
  return format
end)

-- https://www.codegrepper.com/code-examples/lua/lua+split+string+by+delimiter
function Split(s, delimiter)
    result = {};
    for match in (s..delimiter):gmatch("(.-)"..delimiter) do
        table.insert(result, match);
    end
    return result;
end

wezterm.on("bell", function(window, pane)
  local lastOutput = tonumber( pane:get_user_vars().lastOutput )
  if lastOutput ~= nil then
    local lines = Split( pane:get_logical_lines_as_text(), "\n" )
    lastLines[ pane:pane_id() ] = lines[ #lines ]
  end
end)

return {}
```

**Reply by wez** [reply]
Thanks for sharing this!  You motivated me to hook something up in wezterm itself.
https://github.com/wez/wezterm/commit/0bf1b8d13eb4c1ec132d609e12ae7ed78266688b allows for this:

```lua
local wezterm = require 'wezterm'

wezterm.on("format-tab-title", function(tab, tabs, panes, config, hover, max_width)
  if tab.is_active then
    return {
      {Background={Color="blue"}},
      {Text=" " .. tab.active_pane.title .. " "},
    }
  end
  local has_unseen_output = false
  for _, pane in ipairs(tab.panes) do
    if pane.has_unseen_output then
      has_unseen_output = true
      break;
    end
  end
  if has_unseen_output then
    return {
      {Background={Color="Orange"}},
      {Text=" " .. tab.active_pane.title .. " "},
    }
  end
  return tab.active_pane.title
end)

return {
}
```

I need to write up docs around this but have to run now; I'll come back to it later.

Briefly, things to note:

* First, just wanted to note that spawning a new tab doesn't tell the old tab it lost focus, but switching back and forth works fine.  Will fix this later.
* I added `has_unseen_output` field to `PaneInformation` and a `pane:has_unseen_output()` method to tell if there has been output in a pane since it last lost focus.
* I added an event that allows unseen output to signal the gui to update the title bar so that `format-tab-title` will be triggered
* Nuanced behavior: `format-tab-title`'s `panes` parameter is only ever the list of panes for the *active* tab, even when the current `tab` is no the active tab.  Will need to think about whether this should be changed to match the current tab in a future release.
* Meanwhile, I added a `panes` field to `TabInformation` to get the pane information list for that tab, so that there is a way to access that info

The example above shows how to iterate the panes in a tab to see if any of them have unseen output.  For an inactive tab with unseen output, the tab color is set to orange.

You should be able to download a nightly build with these changes within an hour of this comment being posted.

**Reply by hoverduck** [reply]
Phenomenal!  I'll check the nightly build out later today :)

**Comment by hoverduck** [comment]
I checked out the nightly build last night and the new tab activity feature works great!  Unfortunately for me I also ran into too many issues with my key maps on Colemak (I assume because of #1483) so I reverted back to the stable release for now.  But I'll switch back to nightly when I have more time to play with it and figure out what's going on :)

**Comment by madhur** [comment]
Are we going to roll out this feature?

**Reply by wez** [reply]
It shipped over a year ago already in https://wezfurlong.org/wezterm/changelog.html#20220319-142410-0fcdea07

**Comment by sandeep-ad** [comment]
Could you please make it look like Tmux where tab name does not change with cwd and where pressing leader key and comma allows to rename tab title.
